java/ql/lib/semmle/code/java/security/InsufficientKeySize.qll

```diff
@@ -0,0 +1,191 @@
+private class AsymmetricEcSink extends InsufficientKeySizeSink {
+  AsymmetricEcSink() {
+    exists(AsymmetricInitMethodAccess ma, AsymmetricKeyGenerator kg |
+      kg.getAlgoName().matches("EC%") and
+      DataFlow::localExprFlow(kg, ma.getQualifier()) and
+      this.asExpr() = ma.getKeySizeArg()
+    )
+    or
+    exists(AsymmetricEcSpec s | this.asExpr() = s.getKeySizeArg())
+  }
+
+  override predicate hasState(DataFlow::FlowState state) {
+    state = getMinAsymEcKeySize().toString()
+  }
+}
+
+/** A sink for an insufficient key size used in AES algorithms. */
+private class SymmetricSink extends InsufficientKeySizeSink {
+  SymmetricSink() {
+    exists(SymmetricInitMethodAccess ma, SymmetricKeyGenerator kg |
+      kg.getAlgoName() = "AES" and
+      DataFlow::localExprFlow(kg, ma.getQualifier()) and
+      this.asExpr() = ma.getKeySizeArg()
+    )
+  }
+
+  override predicate hasState(DataFlow::FlowState state) { state = getMinSymKeySize().toString() }
+}
+
+// ********************** SINKS HELPER CLASSES & PREDICATES **********************
+/** A call to a method that initializes a key generator. */
+abstract private class KeyGenInitMethodAccess extends MethodAccess {
+  /** Gets the `keysize` argument of this call. */
+  Argument getKeySizeArg() { result = this.getArgument(0) }
+}
+
+/**
+ * A call to the `initialize` method declared in `java.security.KeyPairGenerator`
+ * or to the `init` method declared in `java.security.AlgorithmParameterGenerator`.
+ */
+private class AsymmetricInitMethodAccess extends KeyGenInitMethodAccess {
+  AsymmetricInitMethodAccess() {
+    this.getMethod() instanceof KeyPairGeneratorInitMethod or
+    this.getMethod() instanceof AlgoParamGeneratorInitMethod
+  }
+}
+
+/** A call to the `init` method declared in `javax.crypto.KeyGenerator`. */
+private class SymmetricInitMethodAccess extends KeyGenInitMethodAccess {
+  SymmetricInitMethodAccess() { this.getMethod() instanceof KeyGeneratorInitMethod }
+}
+
+/** An instance of a generator that specifies an encryption algorithm. */
+abstract private class AlgoGeneratorObject extends CryptoAlgoSpec {
+  /** Returns an uppercase string representing the algorithm name specified by this generator object. */
+  string getAlgoName() { result = this.getAlgoSpec().(StringLiteral).getValue().toUpperCase() }
+}
+
+/**
+ * An instance of a `java.security.KeyPairGenerator`
+ * or of a `java.security.AlgorithmParameterGenerator`.
+ */
+private class AsymmetricKeyGenerator extends AlgoGeneratorObject {
+  AsymmetricKeyGenerator() {
+    this instanceof JavaSecurityKeyPairGenerator or
+    this instanceof JavaSecurityAlgoParamGenerator
+  }
+
+  override Expr getAlgoSpec() { result = this.getAlgoSpec() }
+}
+
+/** An instance of a `javax.crypto.KeyGenerator`. */
+private class SymmetricKeyGenerator extends AlgoGeneratorObject {
+  SymmetricKeyGenerator() { this instanceof JavaxCryptoKeyGenerator }
+
+  override Expr getAlgoSpec() { result = this.getAlgoSpec() }
```

Comment:
Same as above, but we can be even more explicit in this case if we apply the previous suggestion:

```suggestion
  override Expr getAlgoSpec() { result = JavaxCryptoKeyGenerator.super.getAlgoSpec() }
```